subprojects/core/src/integTest/groovy/org/gradle/api/internal/changedetection/CorruptedTaskHistoryIntegrationTest.groovy

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.changedetection
+
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+import spock.lang.Issue
+
+class CorruptedTaskHistoryIntegrationTest extends AbstractIntegrationSpec {
+
+    // If this test starts to be flaky it points to a problem in the code!
+    // See the linked issue.
+    @Issue("https://github.com/gradle/gradle/issues/2827")
+    def "broken build doesn't corrupt the artifact history"() {
+        def numberOfFiles = 10
+        def numberOfInputProperties = 10
+        def numberOfTasks = 100
+        def millisecondsToKill = 300
+
+        setupTestProject(numberOfFiles, numberOfInputProperties, numberOfTasks, millisecondsToKill)
+
+        executer.beforeExecute {
+            // We need a separate JVM in order not to kill the test JVM
+            requireGradleDistribution()
+        }
+
+        when:
+        succeeds("createFiles")
+        succeeds("clean")
+        fails("createFiles", "-PkillMe=true", "--max-workers=${numberOfTasks}")
+
+        then:
+        file('build/output10').allDescendants().size() == numberOfFiles
```

Comment:
Is this check necessary? Couldnt a loaded agent fail to create all of these files before the daemon is killed?

Reply:
This seems to be one of the first tasks to be executed. I want the build to be killed when it is more or less halfway through. How about I add a check that the total number of files in the build directory is between `#tasks * #filesPerTask * (0.2 - 0.8)`? so 20% to 80% of all the files have been created?

Reply:
I made the change I proposed.